### ISSUE TITLE
Automatically validate produced HTML

### ISSUE THREAD

**Issue by martinmo** [st-tu-dresden/inloop#246]
With https://validator.w3.org/nu/, there is a good open source HTML5 validator that can be used to check HTML files and URLs from the command line. We should integrate it into our CI. This requires a little work because we cannot simply scan the Django templates.

**Comment by PhilippMatthes** [comment]
I would address this issue after https://github.com/st-tu-dresden/inloop/issues/245 so that the html can be validated in our integration tests. I'd use [this Python library derived from the validator you supplied](https://github.com/svenkreiss/html5validator) @martinmo. What do you think about this?

**Comment by martinmo** [comment]
Did you check whether the nu validator itself has a command line interface? Or in other words, why do we need to use the Python bindings?

**Comment by PhilippMatthes** [comment]
@martinmo the nu validator has a command line interface. But as you said, we cannot simply check our templates, because they contain Django's renderable template tags.
So my approach would be to design integration tests first in #245 and then partially use these to integrate our template checks, preferably using the derived Python library. 
Do you see any workaround with which we would be able to feed in our templates without rendering them in the first place?

**Comment by martinmo** [comment]
I would do it as follows: require that the development server is started (i.e., `honcho start`). Find a way that the program can login to see the protected pages. Then check a set of predefined URLs on localhost:8000.

**Comment by PhilippMatthes** [comment]
@martinmo am I interpreting this correctly, that you want to run the html validation in a shell script? This is possible although sketchy as of my opinion. Although it is possible to obtain the CSRF middleware token and reuse it for html validation of restricted pages, we have several problems with this approach, that I have been confronted with during experimenting with your suggested approach:

1. It is required, as you said, to run the development server first. Though notably, the command `honcho start` blocks any further action after the server is started. Alternatively, one can use the `honcho start & ...` bash syntax to run both jobs in parallel, where `...` stands for all further verification jobs, but then the verification jobs don't wait for the server to be started up and most likely finish before the server started. Using a delay here is a pitfall, and I see no other option since there are no e.g. signals passed from `honcho start` whatsoever, that could be used to control this flow.
2. Running tests on predefined urls introduces more constraints on url mappings, where changing an url might result in test failures.
 
Both problems can be worked around by using python bindings as discussed earlier.

**Comment by martinmo** [comment]
I see what you mean. It doesn't necessarily need to be a shell script, especially when there will be several processes to run in parallel. However, it is not part of the unit test suite. For me it belongs into the category of linters, and my goal is that it should be as simple as possible. Even if we start out with a static hard-coded list of URLs to check at the beginning, this would still be better than having no checks at all.

**Comment by PhilippMatthes** [comment]
For efficiency, this issue should be further discussed in a meeting.

**Comment by martinmo** [comment]
In the meantime I experimented a bit with vnu.jar and also took a deeper look at the library you proposed. TL;DR is that I couldn't find a satisfying solution in a reasonable amount of time. Especially, it's cumbersome to validate pages that require authentication.

However, for semi-automatic validation using the "nu" validator, I found https://pypi.org/project/django-html-validator/. This should be more than enough to catch HTML bugs during PR reviews, for example. (Because the validator  starts up very slowly, it is better to use the local webserver mode of vnu.jar and set `HTMLVALIDATOR_VNU_URL` in the django settings.)
